### PR TITLE
Removed c-style pointer casts and redundant `reinterpret_cast`s in cudf::io

### LINKED ISSUE #6029: [BUG] reinterpret_cast is misused in many places in cuIO
There are (at least) two ways in which the `reinterpret_cast` is misused:
1. Used instead of `static_cast`:
https://github.com/rapidsai/cudf/blob/f7fbc1160b17b969db2708e0cf6033d3db9fc1cf/cpp/src/io/avro/avro_gpu.cu#L95
static cast should be used to cast from `void*`.
2. The use causes undefined behavior:
https://github.com/rapidsai/cudf/blob/f7fbc1160b17b969db2708e0cf6033d3db9fc1cf/cpp/src/io/avro/avro_gpu.cu#L112
https://github.com/rapidsai/cudf/blob/f7fbc1160b17b969db2708e0cf6033d3db9fc1cf/cpp/src/io/avro/avro_gpu.cu#L131-L132
These uses break the [type aliasing rules](https://en.cppreference.com/w/cpp/language/reinterpret_cast) as they cast between pointers to types that are not _similar_.

Proposed solution:

- Use `static_cast` where appropriate.
- Buffers whose data type is not known at compile time (input/output buffers usually) should be of type void* instead of uint8_t*, as they commonly are now.
  - Addition of a `hostdevice_buffer` type might be beneficial here.
- Where the options above are not applicable, use `std::memcpy` instead of the cast.

TODO list of components:
- [x] Avro
- [ ] Compression
- [ ] CSV
- [x] ORC
- [ ] Parquet
- [ ] Stats
- [ ] Utilities

### PR DESCRIPTION
**Pull request by lamarrr**
This PR is in response to Issue https://github.com/rapidsai/cudf/issues/6029. I removed some redundant uses of `reinterpret_cast` and additionally removing some uses of c-style pointer casts.

Justification for removing c-style pointer casts:
C-style pointer casts don't make a distinction between any of these:
 - `reinterpret_cast`
 - `static_cast`
 - `const_cast`
 - `volatile` cast

this inherently makes it difficult to identify and distinguish between them which is needed in the case where we want to find them, especially whilst `reinterpret_cast` and `const_cast` can potentially be a serious source of bugs.

Fixes #6029

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by harrism** [comment]
> Justification for removing c-style pointer casts

No need for justification here!

**Comment by kkraus14** [comment]
ok to test

File: cpp/src/io/comp/uncomp.cpp [hunk 1]
```diff
@@ -219,7 +220,7 @@ int cpu_inflate(uint8_t *uncomp_data, size_t *destLen, const uint8_t *comp_data,
   z_stream strm;
 
   memset(&strm, 0, sizeof(strm));
-  strm.next_in   = (Bytef *)comp_data;
+  strm.next_in   = const_cast<Bytef *>(reinterpret_cast<Bytef const *>(comp_data));
```
Comment (vuule) [inline cpp/src/io/comp/uncomp.cpp:223]: Wow, a very smelly code hidden behind a cast here. 
Defining ZLIB_CONST would make `next_in` const, but I don't know if it affects zlib performance. I think we can instead pass a non-const `comp_data` pointer here (note for a separate PR).

File: cpp/src/partitioning/partitioning.cu [hunk 2]
```diff
@@ -257,8 +257,7 @@ __global__ void copy_block_partitions(InputIter input_iter,
   auto block_output = reinterpret_cast<DataType*>(shared_memory);
   auto partition_offset_shared =
     reinterpret_cast<size_type*>(block_output + OPTIMIZED_BLOCK_SIZE * OPTIMIZED_ROWS_PER_THREAD);
-  auto partition_offset_global =
-    reinterpret_cast<size_type*>(partition_offset_shared + num_partitions + 1);
+  size_type* partition_offset_global = partition_offset_shared + num_partitions + 1;
```
Comment (vuule) [inline cpp/src/partitioning/partitioning.cu:260]: I'd keep `auto` here. IMO it's easy to see the type from the context.
```suggestion
  auto partition_offset_global = partition_offset_shared + num_partitions + 1;
```

**Review by vuule — CHANGES_REQUESTED** [review]
Changes look good. 
Just got a nitpick and a suggestion to further improve the code :)

File: cpp/src/io/comp/debrotli.cu [hunk 3]
```diff
@@ -1377,12 +1377,12 @@ static __device__ debrotli_huff_tree_group_s *HuffmanTreeGroupInit(debrotli_stat
...
-    s, sizeof(debrotli_huff_tree_group_s) + ntrees * sizeof(uint16_t *) - sizeof(uint16_t *));
-  group->alphabet_size = (uint16_t)alphabet_size;
-  group->max_symbol    = (uint16_t)max_symbol;
-  group->num_htrees    = (uint16_t)ntrees;
-  group->htrees[0]     = nullptr;
+  debrotli_huff_tree_group_s *group = reinterpret_cast<debrotli_huff_tree_group_s *>(local_alloc(
+    s, sizeof(debrotli_huff_tree_group_s) + ntrees * sizeof(uint16_t *) - sizeof(uint16_t *)));
+  group->alphabet_size              = (uint16_t)alphabet_size;
+  group->max_symbol                 = (uint16_t)max_symbol;
+  group->num_htrees                 = (uint16_t)ntrees;
```
Comment (harrism) [inline cpp/src/io/comp/debrotli.cu:1384]: What about these  c-style casts?  I wonder why it was considered safe to convert  32-bit to 16-bit here.

**Review by harrism — APPROVED** [review]
Thanks for this. Just one question.

**Comment by harrism** [comment]
@lamarrr code freeze for RAPIDS v0.16 starts Oct 8, which means there is less than 48 hours left to get this PR finished. Will you have time?

File: cpp/src/io/comp/cpu_unbz2.cpp [hunk 4]
```diff
@@ -568,11 +569,11 @@ int32_t cpu_bz2_uncompress(
       s.cur    = source + (size_t)(bit_offs >> 3);
       s.bitpos = (uint32_t)(bit_offs & 7);
       if (s.cur + 8 > s.end) return BZ_PARAM_ERROR;
-      s.bitbuf = bswap_64(*(const uint64_t *)s.cur);
+      s.bitbuf = bswap_64(*reinterpret_cast<const uint64_t *>(s.cur));
     }
   }
 
-  s.tt = (uint32_t *)malloc(s.blockSize100k * 100000 * sizeof(int32_t));
+  s.tt = static_cast<uint32_t *>(malloc(s.blockSize100k * 100000 * sizeof(int32_t)));
```
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/src/io/comp/cpu_unbz2.cpp:576]: Alright, I will. Thanks for the suggestion!

File: cpp/src/io/comp/uncomp.cpp (hunk 1 above)
Comment (lamarrr) [reply cpp/src/io/comp/uncomp.cpp:226]: Smelly indeed.
I intentionally left it like that as we have to account for both the scenario where `ZLIB_CONST` is defined as `const` and not.
The other contexts where the function `cpu_inflate` is used requires that it is `const` (i.e. they pass `const` arguments). So we'd have to perform the `const_cast` anyway.
I think the best solution would be to remove `Z_CONST`, but I'm not sure how that'd affect the output and performance as it'd require doing more investigation as I'd rather not make assumptions.

File: cpp/src/partitioning/partitioning.cu (hunk 2 above)
Comment (lamarrr) [reply cpp/src/partitioning/partitioning.cu]: I've resolved this in the commit below. Thanks!

File: cpp/src/io/comp/cpu_unbz2.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/src/io/comp/cpu_unbz2.cpp]: I've made the change here (https://github.com/rapidsai/cudf/pull/6386/commits/59d440149408608adc59e0bbb5db6efb01e7c289). Thanks!

**Comment by lamarrr** [comment]
> @lamarrr code freeze for RAPIDS v0.16 starts Oct 8, which means there is less than 48 hours left to get this PR finished. Will you have time?

Yeah, sure! I'll try to finish it ASAP

**Review by vuule — APPROVED** [review]

File: cpp/src/io/comp/debrotli.cu (hunk 3 above)
Comment (lamarrr) [reply cpp/src/io/comp/debrotli.cu:1384]: I'm just as curious about the motive. There are still a number of c-style casts within `cudf` that need fixing.
This PR addresses the c-style pointer casts in `cudf::io`. If time permits, I'll try to address the c-style value casts too in another PR.

**Comment by lamarrr** [comment]
I’ve made all the necessary changes since 3 days ago @harrism . It seems there are some new conflicts.

**Comment by harrism** [comment]
You can resolve conflicts by merging branch-0.16 into your branch, resolving, commiting and pushing.

**Pull request merged**